server/routes/api/protein.py

```diff
@@ -13,22 +13,411 @@
+        except:
+            continue
+    return DEFAULT_INTERACTION_SCORE
+
+
+def _interactors(interaction_id_or_dcid):
+    '''
+    Parse and return ids of two interactors from interaction DCID of the form 'bio/{protein1 id}_{protein2 id}'
+    E.g. 'bio/P53_HUMAN_CBP_HUMAN' --> 'P53_HUMAN', 'CBP_HUMAN'
+    '''
+    interaction_id = _id(interaction_id_or_dcid)
+    try:
+        protein1, species1, protein2, species2 = interaction_id.split('_')
+    except ValueError as error:
+        _log_ppi(f'Invalid protein identifier "{interaction_id_or_dcid}"',
+                 '_interactors')
+        raise error
+    return {
+        'first_interactor': f'{protein1}_{species1}',
+        'second_interactor': f'{protein2}_{species2}'
+    }
+
+
+def _other_interactor(interaction_dcid, interactor_id):
+    '''
+    Given interaction DCID and the id of one of the interactors, return the id of the other one.
+    E.g. 'bio/P53_HUMAN_CBP_HUMAN', 'P53_HUMAN' --> 'CBP_HUMAN'
+    '''
+    interaction_id = _id(interaction_dcid)
+    interactor_id1, interactor_id2 = _interactors(interaction_id).values()
+    if interactor_id1 == interactor_id:
+        return interactor_id2
+    elif interactor_id2 == interactor_id:
+        return interactor_id1
+    raise ValueError(f'{interactor_id} not in {interaction_dcid}')
+
+
+def _reverse_interaction_id(interaction_id):
+    '''
+    Given interaction ID of the form '{protein1 id}_{protein2 id}', return '{protein2 id}_{protein1 id}'.
+    E.g. 'P53_HUMAN_CBP_HUMAN' --> 'CBP_HUMAN_P53_HUMAN'
+    '''
+    interactors = _interactors(interaction_id)
+    return f'{interactors["second_interactor"]}_{interactors["first_interactor"]}'
+
+
+def _reverse_interaction_dcid(interaction_dcid):
+    '''
+    Given interaction DCID of the form 'bio/{protein1 id}_{protein2 id}', return 'bio/{protein2 id}_{protein1 id}'.
+    E.g. 'bio/P53_HUMAN_CBP_HUMAN' --> 'bio/CBP_HUMAN_P53_HUMAN'
+    '''
+    return _dcid(_reverse_interaction_id(_id(interaction_dcid)))
+
+
+def _filter_interaction_dcids(interaction_dcids):
+    '''
+    Given list of interaction DCIDs, remove self-interactions and deduplicate (keep one copy of each set of duplicates).
+    Self-interactions are interactions DCIDs of the form 'bio/{protein id}_{protein id}' (e.g. 'bio/P53_HUMAN_P53_HUMAN')
+    Duplicates include both exact duplicates and reverse duplicates.
+    E.g. 'bio/P53_HUMAN_CBP_HUMAN' is duplicate with both itself and 'bio/CBP_HUMAN_P53_HUMAN'.
+    '''
+    dcid_set = set()
+    uniques = []
+    for interaction_dcid in interaction_dcids:
+        try:
+            protein_id1, protein_id2 = _interactors(interaction_dcid).values()
+        # skip malformatted ids
+        except ValueError:
+            continue
+        # filter out self-interactions
+        if protein_id1 == protein_id2:
+            continue
+        reverse_dcid = _reverse_interaction_dcid(interaction_dcid)
```

Comment:
I think this should also be wrapped in a try because _reverse_interaction_dcid could throw an error, but I don't think this filter method should end up throwing any errors ..

Reply:
Oh good catch will do!